File: detection_ws/src/value_map/value_map/value_map.py
from rclpy.lifecycle import LifecycleNode
from rclpy.lifecycle import State
import rclpy
from rclpy.lifecycle import TransitionCallbackReturn
from rcl_interfaces.msg import ParameterDescriptor
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
from visualization_msgs.msg import Marker
from value_map.service_handler import ServiceHandler
from tf2_ros import Buffer, TransformListener
from tf2_ros import LookupException, ConnectivityException, ExtrapolationException
from geometry_msgs.msg import Pose
from value_map.semantic_value_map import SemanticValueMap
import time

class ValueMap(LifecycleNode):
    def __init__(self):
        super().__init__('value_map_node')
        self.get_logger().info("Initializing Value Map Node...")

        self.declare_parameter('timer_frequency', 5.0, ParameterDescriptor(description='Frequency of the timer.'))
        self.timer_frequency = self.get_parameter("timer_frequency").get_parameter_value().double_value

        self.timer = None
        self.bridge = CvBridge()

        self.text_query = "brown chair"
        self.rgb_image = None
        self.map = None

        self.marker_pub = self.create_publisher(Marker, "/semantic_score_marker", 1)
        self.service_handler = None
        self.get_logger().info("Finished initializing Value Map Node")

    def on_configure(self, state):
        self.get_logger().info('Configuring Value Map Node...')

        self.tf_buffer = Buffer()
        self.tf_listener = TransformListener(self.tf_buffer, self)

        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: State):
        self.get_logger().info('Activating...')

        try:
            # Subscribers
            self.image_sub = self.create_subscription(Image, '/rgb', self.image_callback, 10)

            # Initialize service handler with this node
            self.service_handler = ServiceHandler(self)

            self.semantic_map = SemanticValueMap(self)

            self.createTimer()

            self.get_logger().info('Value Map Node activated.')
            return TransitionCallbackReturn.SUCCESS

        except Exception as e:
            self.get_logger().error(f'Activation failed: {e}')
            return TransitionCallbackReturn.FAILURE

    def on_cleanup(self, state: State):
        self.get_logger().info('Cleaning up...')
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state: State):
        self.get_logger().info('Shutting down...')
        return TransitionCallbackReturn.SUCCESS

    def createTimer(self):
        self.timer = self.create_timer(1.0 / self.timer_frequency, self.timer_callback)
        self.get_logger().info(f"Timer started with frequency: {self.timer_frequency} Hz")

    def image_callback(self, msg):
        self.rgb_image = msg

    def timer_callback(self):
        start_time = time.time()  # Startzeit messen

        if self.rgb_image is None:
            self.get_logger().warn("No RGB image available.")
            return

        current_pose = self.get_pose()

        if current_pose is None:
            self.get_logger().warn("Current pose could not be retrieved.")
            return

        try:
            panoptic_segmented_image = self.service_handler.call_panoptic(self.rgb_image)
            object_segmented_image = self.service_handler.call_object_segmentation(self.rgb_image, self.text_query)
            semantic_similarity_score = self.service_handler.call_semantic_similarity(self.rgb_image, self.text_query)
            
        except Exception as e:
            self.get_logger().error(f"Service calls failed: {e}")
            # Check VLM namespace
            self.get_logger().warn("Check if the VLM Server namespace is correctly set up.")
            self.get_logger().warn("Current VLM namespace: " + self.service_handler.vlm_namespace)
            return

        if semantic_similarity_score is None:
            self.get_logger().warn("No semantic similarity score received.")
            return
        
        self.get_logger().info(f"Text query: {self.text_query}")
        self.get_logger().info(f"Received semantic similarity score: {semantic_similarity_score}")

        self.semantic_map.update_semantic_map(semantic_similarity_score, current_pose)

        # --- Messung der Ausführungsdauer ---
        elapsed = time.time() - start_time
        max_allowed = 1.0 / self.timer_frequency

        if elapsed > max_allowed:
            self.get_logger().warn(f"Timer callback took {elapsed:.3f}s, which exceeds the target {max_allowed:.3f}s.")
        else:
            self.get_logger().info(f"Timer callback duration: {elapsed:.3f}s")

    def get_pose(self):
        try:
            transform = self.tf_buffer.lookup_transform("map", "base_link", rclpy.time.Time())
            translation = transform.transform.translation
            rotation = transform.transform.rotation

            pose = Pose()
            pose.position.x = translation.x
            pose.position.y = translation.y
            pose.position.z = translation.z
            pose.orientation.x = rotation.x
            pose.orientation.y = rotation.y
            pose.orientation.z = rotation.z
            pose.orientation.w = rotation.w
            return pose
        except (LookupException, ConnectivityException, ExtrapolationException) as e:
            self.get_logger().warn(f"TF transform failed: {e}")
            return None

File: detection_ws/src/value_map/src/value_map/value_map.cpp
#include "value_map.hpp"

ValueMap::ValueMap(const rclcpp::NodeOptions & options)
: LifecycleNode("value_map_node", options)
{
    RCLCPP_INFO(this->get_logger(), "%s[Lifecycle]%s ValueMap Node %sinitialized%s.",
                BLUE, RESET, GREEN, RESET);

    // Declare parameters
    declare_parameter<std::string>("vlm_target_namespace", "/blip2_ros");

    semanticMap = std::make_unique<SemanticValueMap>(this);
    serviceHandler = std::make_unique<ServiceHandler>(this);
    robot = std::make_unique<Robot>(this);
}

ValueMap::~ValueMap() 
{
    RCLCPP_INFO(this->get_logger(), "%s[Lifecycle]%s ValueMap Node %sdestructed%s.",
                BLUE, RESET, YELLOW, RESET);
}

CallbackReturn ValueMap::on_configure(const rclcpp_lifecycle::State &)
{
    RCLCPP_INFO(this->get_logger(), "%s[Transition]%s Configuring ValueMap...", BLUE, RESET);

    this->declare_parameter<double>("timer_frequency", 5.0);
    this->declare_parameter<std::string>("semantic_prompt_topic", "/user_prompt");

    this->get_parameter("timer_frequency", timerFrequency);
    this->get_parameter("semantic_prompt_topic", semanticPromptTopic);

    if(!semanticMap->on_configure()) 
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sconfigure%s Node.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    if(!serviceHandler->on_configure())
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sconfigure%s ServiceHandler.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    if(!robot->on_configure())
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sconfigure%s Robot.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    semanticPromptSub = this->create_subscription<multimodal_query_msgs::msg::SemanticPrompt>(
        semanticPromptTopic, 10, std::bind(&ValueMap::semanticPromptCallback, this, std::placeholders::_1));

    cosineSimPub = this->create_publisher<std_msgs::msg::Float32>("cosine_similarity", 10);

    return CallbackReturn::SUCCESS;

}

CallbackReturn ValueMap::on_activate(const rclcpp_lifecycle::State &)
{
    RCLCPP_INFO(this->get_logger(), "%s[Transition]%s Activating ValueMap...", BLUE, RESET);

    auto interval = std::chrono::duration<double>(1.0 / timerFrequency);
    RCLCPP_INFO(this->get_logger(), "Publishing with a time interval of %.3f seconds", interval.count());
    timer = this->create_wall_timer(
        std::chrono::duration_cast<std::chrono::milliseconds>(interval),    
        std::bind(&ValueMap::timerCallback, this)
    );

    if (!semanticMap->on_activate()) 
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sactivate%s Node.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    if (!serviceHandler->on_activate())
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sactivate%s ServiceHandler.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    if (!robot->on_activate())
    {
        RCLCPP_INFO(this->get_logger(), "%s[Status]%s Could not %sactivate%s Robot.", BLUE, RESET, RED, RESET);
        return CallbackReturn::FAILURE;
    }

    return CallbackReturn::SUCCESS;
}

CallbackReturn ValueMap::on_deactivate(const rclcpp_lifecycle::State &)
{
    RCLCPP_INFO(this->get_logger(), "%s[Transition]%s Deactivating ValueMap...", BLUE, RESET);
    semanticMap->on_deactivate();
    serviceHandler->on_deactivate();
    robot->on_deactivate();
    cosineSimPub.reset();
    semanticPromptSub.reset();
    return CallbackReturn::SUCCESS;
}

CallbackReturn ValueMap::on_cleanup(const rclcpp_lifecycle::State &)
{
    RCLCPP_INFO(this->get_logger(), "%s[Transition]%s Cleaning up ValueMap...", BLUE, RESET);
    if (timer){ timer.reset();}

    semanticMap->on_cleanup();
    serviceHandler->on_cleanup();
    robot->on_cleanup();
    cosineSimPub.reset();
    semanticPromptSub.reset();

    RCLCPP_INFO(this->get_logger(), "%s[Cleanup]%s Timer and subscriber cleaned up.", BLUE, RESET);
    return CallbackReturn::SUCCESS;
}

CallbackReturn ValueMap::on_shutdown(const rclcpp_lifecycle::State &)
{
    RCLCPP_INFO(this->get_logger(), "%s[Transition]%s Shutting down ValueMap...", BLUE, RESET);
    semanticMap->on_shutdown();
    serviceHandler->on_shutdown();
    robot->on_shutdown();
    cosineSimPub.reset();
    semanticPromptSub.reset();
    return CallbackReturn::SUCCESS;
}

void ValueMap::timerCallback()
{
    /////////////////////////////////////////////////////////////
    // Test Case. Remove when done. Setting hardcoded text prompt.
    // textPrompt = "door";
    /////////////////////////////////////////////////////////////

    const std::string prompt = textPrompt;  // or guard with a mutex if multithreaded
    if (prompt.find_first_not_of(" \t\n\r") == std::string::npos) 
    {
        RCLCPP_INFO(this->get_logger(), "No text prompt received yet.");
        return;
    }

    RCLCPP_DEBUG(this->get_logger(), "Current text prompt: %s", prompt.c_str());

    // Get image
    sensor_msgs::msg::Image::SharedPtr rgbImage = robot->getImage();
    
    // If no image is available, return
    if (!rgbImage) return;

    // Get cosine similarity score
    double score = serviceHandler->getSemanticSimilarityScore(*rgbImage, textPrompt);
    score = std::max(0.0, score);  // Ensure non-negative score
    score = std::min(1.0, score);  // Cap at 1.0

    score = 1.0;        // TEMPORARY OVERRIDE. REMOVE AFTER TESTING.
    
    // Publish cosine similarity score
    publishCosineSimilarity(score);

    // Get pose
    geometry_msgs::msg::PoseStamped::SharedPtr pose = robot->getPose(rgbImage->header.stamp);

    // If pose is empty, return
    if (!pose) return;

    semScore.setScore(score);
    semScore.setHeader(rgbImage->header);

    RCLCPP_DEBUG(this->get_logger(), "%s%s Semantic Similarity: %.2f", BLUE, RESET, score);

    // Update semantic map
    semanticMap->updateSemanticMap(semScore, *pose);

}

void ValueMap::semanticPromptCallback(const multimodal_query_msgs::msg::SemanticPrompt::SharedPtr msg)
{
    RCLCPP_INFO(this->get_logger(), "Received Semantic Prompt: %s", msg->text_query.c_str());
    textPrompt = msg->text_query;
    semanticMap->clearValueMap();
}

void ValueMap::publishCosineSimilarity(double score)
{
    auto message = std_msgs::msg::Float32();
    message.data = static_cast<float>(score);
    cosineSimPub->publish(message);
}